File: backend/src/main/java/com/checkba/service/evidence/AnchorHash.java
package com.checkba.service.evidence;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.text.Normalizer;
import java.util.Map;

/**
 * 锚点文字归一化 + sha256。
 *
 * 规则（前后端共用，改一处必须同步另一处并更新向量 fixtures/anchor-hash-vectors.json）：
 * NFKC → 删除全部 Unicode 空白（含 U+3000）→ 中文标点映射（，。；：！？（））→ 引号一律删 → 保留《》。
 * 前端实现：frontend/src/utils/anchorHash.js。
 *
 * 注意 NFKC 已把全角逗号 U+FF0C 归成 ','，所以 PUNCT 里的 '，' 分支基本不命中；
 * 句号 U+3002 NFKC 不变，必须靠表——向量第 1 条就是为了钉这个。
 */
public final class AnchorHash {

    private static final Map<Character, String> PUNCT = Map.ofEntries(
            Map.entry('，', ","), Map.entry('。', "."), Map.entry('；', ";"), Map.entry('：', ":"),
            Map.entry('！', "!"), Map.entry('？', "?"), Map.entry('（', "("), Map.entry('）', ")"),
            Map.entry('「', ""), Map.entry('」', ""), Map.entry('『', ""), Map.entry('』', ""),
            Map.entry('“', ""), Map.entry('”', ""), Map.entry('‘', ""), Map.entry('’', ""),
            Map.entry('"', ""), Map.entry('\'', ""));

    private AnchorHash() {}

    public static String normalize(String s) {
        if (s == null) return "";
        String n = Normalizer.normalize(s, Normalizer.Form.NFKC);
        StringBuilder b = new StringBuilder(n.length());
        for (int i = 0; i < n.length(); i++) {
            char c = n.charAt(i);
            if (Character.isWhitespace(c) || c == '　') continue;
            String m = PUNCT.get(c);
            if (m != null) b.append(m); else b.append(c);
        }
        return b.toString();
    }

    /** 64 位小写 hex。 */
    public static String of(String s) {
        try {
            byte[] d = MessageDigest.getInstance("SHA-256").digest(normalize(s).getBytes(StandardCharsets.UTF_8));
            StringBuilder h = new StringBuilder(64);
            for (byte x : d) h.append(String.format("%02x", x));
            return h.toString();
        } catch (Exception e) {
            throw new IllegalStateException(e);
        }
    }
}

File: frontend/src/utils/anchorHash.js
// anchorHash.js — EvidenceLink 锚点文字归一化 + sha256（前端份）。
//
// 与后端 backend/src/main/java/com/checkba/service/evidence/AnchorHash.java 是
// **同一算法**：NFKC → 删全部空白（含 U+3000）→ 中文标点映射半角 → 引号一律删
// → 保留《》→ sha256 小写 hex。改一处必须同步另一处并更新两份向量
// tests/evidence/anchor-hash-vectors.json（与后端 fixtures 字节一致）。
//
// NFKC 已把 ，（）：；！？ 归半角；句号 。 不归、靠表（向量第 1 条钉这个）。

const PUNCT = {
  '，': ',', '。': '.', '；': ';', '：': ':', '！': '!', '？': '?', '（': '(', '）': ')',
  '「': '', '」': '', '『': '', '』': '', '“': '', '”': '', '‘': '', '’': '', '"': '', "'": '',
}

export function normalizeAnchor(s) {
  if (s == null) return ''
  const n = String(s).normalize('NFKC')
  let out = ''
  for (const ch of n) {
    if (/\s/.test(ch) || ch === '　') continue
    const m = PUNCT[ch]
    out += m === undefined ? ch : m
  }
  return out
}

/** 64 位小写 hex，Promise（浏览器与 node 18+ 都有 globalThis.crypto.subtle）。 */
export async function anchorHash(s) {
  const bytes = new TextEncoder().encode(normalizeAnchor(s))
  const digest = await globalThis.crypto.subtle.digest('SHA-256', bytes)
  return Array.from(new Uint8Array(digest)).map((b) => b.toString(16).padStart(2, '0')).join('')
}
